src/Compilers/CSharp/Portable/Binder/Binder_TupleOperators.cs

```diff
@@ -0,0 +1,386 @@
+            Debug.Assert(length == right.Length);
+
+            var operatorsBuilder = ArrayBuilder<TupleBinaryOperatorInfo>.GetInstance(length);
+
+            for (int i = 0; i < length; i++)
+            {
+                operatorsBuilder.Add(BindTupleBinaryOperatorInfo(node, kind, left[i], right[i], diagnostics));
+            }
+
+            var compilation = this.Compilation;
+            TypeSymbol leftTupleType = MakeConvertedType(operatorsBuilder, node.Left, left, nullable, compilation, diagnostics, isRight: false);
+            TypeSymbol rightTupleType = MakeConvertedType(operatorsBuilder, node.Right, right, nullable, compilation, diagnostics, isRight: true);
+
+            return new TupleBinaryOperatorInfo.Nested(operatorsBuilder.ToImmutableAndFree(), leftTupleType, rightTupleType);
+        }
+
+        private static int? GetTupleCardinality(BoundExpression expr)
+        {
+            if (expr.Kind == BoundKind.TupleLiteral)
+            {
+                var tuple = (BoundTupleLiteral)expr;
+                return tuple.Arguments.Length;
+            }
+
+            TypeSymbol type = expr.Type;
+            if (type is null)
+            {
+                return null;
+            }
+
+            type = type.StrippedType();
+
+            if (type.IsTupleType)
+            {
+                return type.TupleElementTypes.Length;
+            }
+
+            return null;
+        }
+
+        private static ImmutableArray<BoundExpression> GetTupleArgumentsOrPlaceholders(BoundExpression expr)
+        {
+            if (expr.Kind == BoundKind.TupleLiteral)
+            {
+                return ((BoundTupleLiteral)expr).Arguments;
+            }
+
+            // placeholder bound nodes with the proper types are sufficient to bind the element-wise binary operators
+            return expr.Type.StrippedType().TupleElementTypes
+                .SelectAsArray(t => (BoundExpression)new BoundTupleOperandPlaceholder(expr.Syntax, t));
+        }
+
+        /// <summary>
+        /// Make a tuple type (with appropriate nesting) from the types (on the left or on the right) collected
+        /// from binding element-wise binary operators.
+        /// </summary>
+        private TypeSymbol MakeConvertedType(ArrayBuilder<TupleBinaryOperatorInfo> operators, CSharpSyntaxNode syntax,
+            ImmutableArray<BoundExpression> elements, bool nullable, CSharpCompilation compilation, DiagnosticBag diagnostics, bool isRight)
+        {
+            ImmutableArray<TypeSymbol> convertedTypes = operators.SelectAsArray(o => o.GetConvertedType(isRight));
+            ImmutableArray<Location> elementLocations = elements.SelectAsArray(e => e.Syntax.Location);
+
+            var tuple = TupleTypeSymbol.Create(locationOpt: null, elementTypes: convertedTypes,
+                elementLocations, elementNames: default, compilation,
+                shouldCheckConstraints: true, errorPositions: default, syntax, diagnostics);
```

Comment:
Do we require `System.ValueTuple` for tuple literal equality? `if ((x, y) == (z, w)) { ... }`

Reply:
Looks like we do. I'll add a test and a PROTOTYPE marker to consider relaxing that. My guess is that restriction will have to stay for 7.3.

---
In reply to: [167702107](https://github.com/dotnet/roslyn/pull/23356#discussion_r167702107) [](ancestors = 167702107)